crates/net/network/src/session/mod.rs

```diff
@@ -319,6 +323,23 @@ impl SessionManager {
         });
     }
 
+    pub(crate) fn handle_disconnect_incoming_connection(
+        &mut self,
+        stream: TcpStream,
+        reason: DisconnectReason,
+        counter: &Counter,
+    ) {
+        let secret_key = self.secret_key;
+        let counter_arc = Arc::clone(&counter.0); // Clone the Arc
+
```

Comment:
here we clone `self.graceful_disconnects`, check how many `strong_counters` we have, if this exceeds the MAX_ value we simply drop/return, otherwise we spawn and gracefully disconnect